Source/WebCore/rendering/RenderBox.cpp

```diff
@@ -1434,41 +1434,50 @@ void RenderBox::clearTrimmedMarginsMarkings()
     ensureRareData().setTrimmedMargins(0);
 }
 
-bool RenderBox::hasTrimmedMargin(std::optional<MarginTrimType> marginTrimType) const
+bool RenderBox::canCheckForTrimmedMargin(std::optional<MarginTrimType> marginTrimType) const
 {
-    if (!isInFlow())
+    auto* containingBlock = this->containingBlock();
+    if (!containingBlock || containingBlock->isRenderView())
         return false;
-#if ASSERT_ENABLED
+
+    // Return false for layout type and margin trim combinations that currently do not set the margin
+    // trim rare data in RenderBox
     // containingBlock->isBlockContainer() can return true even if the item is in a RenderFlexibleBox
     // (e.g. buttons) so we should explicitly check that the item is not a flex item to catch block containers here
-    if (auto* containingBlock = this->containingBlock(); containingBlock && !containingBlock->isFlexibleBox() &&  (containingBlock->isBlockContainer() || containingBlock->isRenderGrid())) {
-        ASSERT_NOT_IMPLEMENTED_YET();
+    if (!isFlexItem() && (containingBlock->isRenderGrid() || containingBlock->isBlockContainer())) {
         return false;
     }
-#endif
-    if (!hasRareData())
-        return false;
-    return marginTrimType ? (rareData().trimmedMargins() & static_cast<unsigned>(*marginTrimType)) : rareData().trimmedMargins();
+
+    if (isFlexItem() && (!marginTrimType || marginTrimType == MarginTrimType::BlockStart))
+        return marginTrimType ? containingBlock->style().marginTrim().contains(*marginTrimType) : !containingBlock->style().marginTrim().isEmpty();
+    return false;
 }
 
-bool RenderBox::hasTrimmedMargin(PhysicalDirection physicalDirection) const
+bool RenderBox::hasTrimmedMargin(std::variant<std::monostate, PhysicalDirection, MarginTrimType> direction) const
```

Comment:
This is getting a bit too difficult to parse. Are you sure we need variant to solve this problem?

Reply:
Nope. It would have been nice to get rid of the function overloading imo but not if it worsens the readability of the code. Went back to my original approach and am now just adding the extra check before calling `box->hasTrimmedMargin`

It would be nice to do some clean up of the code, but that might be better in a separate patch when all of the similar logic for each side has been implemented (at least for flex layout)

Also I think this would be a good opportunity to do the function renaming like you mentioned in https://github.com/WebKit/WebKit/pull/11741

The new name is a little verbose, but hopefully it gets the point across and is accurate in its name